flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -207,7 +210,14 @@ public Buffer requestBuffer() throws IOException {
 
 	@Override
 	public BufferBuilder requestBufferBuilderBlocking() throws IOException, InterruptedException {
-		return toBufferBuilder(requestMemorySegment(true));
+		final MemorySegment segment = requestMemorySegment(true);
+
+		synchronized (availableMemorySegments) {
```

Comment:
Here we are acquiring the  `availableMemorySegments` for the second time, can we try to avoid it somehow?

Reply:
I actually also concerned for this part while implementing, and I tried to do it inside `LocalBufferPool#requestMemorySegment(boolean isBlocking)` before, but it seems a bit trouble to refactor that process because there are many internal paths to return/exit. Let me think it through again.